File: cpp/Basics/OOPS/PaymentService.cpp
#include <map>
#include <string>
#include "paymentmethod.cpp"

class PaymentService {
private:
    std::map<std::string, PaymentMethod*> paymentMethods;

public:
    PaymentService() {}

    void addPaymentMethod(const std::string& name, PaymentMethod* pm) {
        paymentMethods[name] = pm;
    }

    void makePayment(const std::string& name) {
        std::map<std::string, PaymentMethod*>::iterator it = paymentMethods.find(name);
        if (it != paymentMethods.end()) {
            it->second->pay();
        }
    }

    ~PaymentService() {
        for (std::map<std::string, PaymentMethod*>::iterator it = paymentMethods.begin(); 
             it != paymentMethods.end(); ++it) {
            delete it->second;
        }
    }
};

File: design-patterns/src/main/java/org/prateek/Basics/OOPS/PaymentService.java
package org.Prathamesh.Basics.OOPS;

import java.util.HashMap;

public class PaymentService {
    //Storing + Making payments
    HashMap<String,PaymentMethod> paymentMethods;

    PaymentService(){
        paymentMethods = new HashMap<>();
    }

    public void addPaymentMethod(String name,PaymentMethod pm){
          paymentMethods.put(name,pm);
    }
    public void makePayment(String name){
            PaymentMethod pm = paymentMethods.get(name);
            pm.pay(); //Run Time Polymorphism
    }
}
